doc/src/content/xdocs/spec.xml

```diff
@@ -273,60 +273,125 @@
-	  <li>A fullname is specified.  If the name specified contains
-	  a dot, then it is assumed to be a fullname, and any
-	  namespace also specified is ignored.  For example,
-	  use <code>"name": "org.foo.X"</code> to indicate the
-	  fullname <code>org.foo.X</code>.</li>
-	  <li>A name only is specified, i.e., a name that contains no
-	  dots.  In this case the namespace is taken from the most
-	  tightly enclosing schema or protocol.  For example,
-	  if <code>"name": "X"</code> is specified, and this occurs
-	  within a field of the record definition
-	  of <code>org.foo.Y</code>, then the fullname
-	  is <code>org.foo.X</code>. If there is no enclosing
-	  namespace then the null namespace is used.</li>
-	</ul>
-	<p>References to previously defined names are as in the latter
-	two cases above: if they contain a dot they are a fullname, if
-	they do not contain a dot, the namespace is the namespace of
-	the enclosing definition.</p>
-	<p>Primitive type names have no namespace and their names may
-	not be defined in any namespace.</p>
-	<p> A schema or protocol may not contain multiple definitions
-	of a fullname.  Further, a name must be defined before it is
-	used ("before" in the depth-first, left-to-right traversal of
-	the JSON parse tree, where the <code>types</code> attribute of
-	a protocol is always deemed to come "before" the
-	<code>messages</code> attribute.)
-	</p>
+          determined according to the algorithm below the example:
+        </p>
+        <source>
+{
+  "type": "record",
+  "name": "Example",
+  "doc": "A simple name (attribute) and no namespace attribute: use the null namespace (\"\"); the fullname is 'Example'.",
+  "fields": [
+    {
+      "name": "inheritNull",
+      "type": {
+        "type": "enum",
+        "name": "Simple",
+        "doc": "A simple name (attribute) and no namespace attribute: inherit the null namespace of the enclosing type 'Example'. The fullname is 'Simple'.",
+        "symbols": ["a", "b"]
+      }
+    }, {
+      "name": "explicitNamespace",
+      "type": {
+        "type": "fixed",
+        "name": "Simple",
+        "namespace": "explicit",
+        "doc": "A simple name (attribute) and a namespace (attribute); the fullname is 'explicit.Simple' (this is a different type than of the 'inheritNull' field).",
+        "size": 12
+      }
+    }, {
+      "name": "fullName",
+      "type": {
+        "type": "record",
+        "name": "a.full.Name",
+        "namespace": "ignored",
+        "doc": "A name attribute with a fullname, so the namespace attribute is ignored. The fullname is 'a.full.Name', and the namespace is 'a.full'.",
+        "fields": [
+          {
+            "name": "inheritNamespace",
+            "type": {
+              "type": "enum",
+              "name": "Understanding",
+              "doc": "A simple name (attribute) and no namespace attribute: inherit the namespace of the enclosing type 'a.full.Name'. The fullname is 'a.full.Understanding'.",
```

Comment:
@RyanSkraba I really appreciate this writeup as I'm reimplementing schema parsing for Elixir, and struggling with the inconsistencies of how existing libraries resolve references.

Can you answer a question for me?

Imagine the following schema

```json
{
  "name": "a",
  "namespace": "top",
  "type": "record",
  "fields": [
      {"name": "b", "type": {"name": "c", "type": "string"}}
  ]
}
```

Would the fullnames here be either:

1. `top.a`, `top.b`, and `top.c`
2.  `top.a`, `top.b`, `c`

It's unclear to me how namespaces should propagate. It seems like 1 should be the correct answer, but unclear as most libraries seem to not enforce this

Reply:
Hello!  I think this is a good question, and probably could also be clarified in the specification.

1. Field names don't have namespaces, and
2. Unnamed types such as `string` don't have names and don't inherit namespaces.

So there is only one named type in your example `top.a` and one field name `b`.  The attribute `"name": "c"` on the `string` looks like it is accepted, but silently dropped in the Java SDK.

Where the specification is vague: "Attributes not defined in this document are permitted as metadata, but must not affect the format of serialized data."  I think it makes sense to drop metadata that are applied to primitive schemas, if we consider that the type `string` is an immutable singleton instance in the SDK.

Reply:
I created AVRO-3430 for discussion on whether Java is correct to silently drop the `"name": "c"` attribute on the primitive string.

Reply:
I believe it is correct as you pointed out that it's technically metadata here. The canonical parsing form would drop it down to just the primitive.

Can you clarify where in the spec it specifies that Record fields do not have namespaces and thus do not have full names? Clearly they do not have a namespace field, but that nuance was not obvious.

This all seems like useful data to go back into the spec

Reply:
Also, if you're interested, this is the PR that brought up these questions https://github.com/beam-community/avro_ex/pull/62

Reply:
Another question is that if field names do not have namespaces, how do field aliases come into play?

Reply:
It's all about context.

A namespace is a context for (named) schemata. Named schemata (of types `record`/`error`, `enum` or `fixed`) must have a unique name within their context. This results in globally unique _full names_ (i.e., names prefixed with the namespace and a dot).

Fields cannot share a context with schemata, because they're not schemata. The context where fields are known is their `record` schema. Field names must be unique there.

Aliases are alternate names for schemata and fields, and the same rules as for names apply. Aliases for schemata are interpreted in the namespace of the schema, or globally if the alias includes a namespace (i.e., it's a _full name_). Aliases for fields are interpreted within the record that contains the field (and like field names, cannot reference another context/`record` schema).

Reply:
> Can you clarify where in the spec it specifies that Record fields do not have namespaces and thus do not have full names?

I agree that it's not great, and specified "by omission" [here](https://github.com/apache/avro/blob/release-1.11.0/doc/src/content/xdocs/spec.xml#L277-L282)  To paraphrase: `Record, enum and fixed each have a fullname composed of namespace and name... The name portion of a fullname, record field names, and enum symbols must match [an alphanumeric regex without dot]`  I raised AVRO-3436 because it would be simple to clarify!